File: tests/cpp/unit/testUnitPreprocessor.cpp
#include "preprocessor.hpp"
#include <gtest/gtest.h>

class TestPreprocessor : public ::testing::Test
{
  protected:
    Preprocessor preprocessor;
};

TEST_F(TestPreprocessor, processFile)
{
  fs::path filePath = "../../../resources/jazz/jazz.00001.wav";
  uint numFramesExpected = 1 + preprocessor.config().size / preprocessor.config().hop;
  uint mfccSizeExpected = preprocessor.config().nmfcc;

  torch::Tensor mfcc = preprocessor.processFile(filePath);
  ASSERT_EQ(mfcc.numel(), numFramesExpected * mfccSizeExpected) << "Invalid number of the MFCC frames.";
}

TEST_F(TestPreprocessor, processFileThrowError)
{
  fs::path filePath = "../../../resources/jazz/jazz.00054.txt";
  EXPECT_THROW(preprocessor.processFile(filePath), std::runtime_error);
}

TEST_F(TestPreprocessor, normalizeData)
{
  uint numFrames = 1 + preprocessor.config().size / preprocessor.config().hop;
  uint mfccSize = preprocessor.config().nmfcc;
  torch::Tensor x = 2.0f * torch::ones({numFrames, mfccSize});
  torch::Tensor meanExpected = torch::ones({numFrames, mfccSize});
  torch::Tensor stdExpected = 2.0f * torch::ones({numFrames, mfccSize});
  torch::Tensor yExpected = 0.5f * torch::ones({numFrames, mfccSize});

  preprocessor.setMean(meanExpected);
  preprocessor.setStd(stdExpected);
  torch::Tensor y = preprocessor.normalizeData(x);
  ASSERT_TRUE(torch::equal(preprocessor.mean(), meanExpected)) << "Invalid mean value.";
  ASSERT_TRUE(torch::equal(preprocessor.std(), stdExpected)) << "Invalid standard deviation value.";
  ASSERT_TRUE(torch::equal(y, yExpected)) << "Invalid normalized signal.";
}

File: tests/python/unit/test_unit_preprocessor.py
from preprocessor import Preprocessor
import pytest
import torch

@pytest.fixture
def preprocessor():
  return Preprocessor()

def test_preprocessor_process_file(preprocessor):
  file_path = "./resources/jazz/jazz.00001.wav"
  num_frames_expected = 1 + int(preprocessor.config.size / preprocessor.config.hop)
  mfcc_size_expected = preprocessor.config.nmfcc
  mfcc = preprocessor.process_file(file_path)
  assert(mfcc.numel() == num_frames_expected * mfcc_size_expected), "Invalid size of the MFCC."

def test_preprocessor_process_file_exception(preprocessor):
  file_path = "./resources/jazz/jazz.00054.txt"
  with pytest.raises(FileNotFoundError, match=f"Preprocessor.__load_and_crop: File: {file_path} is invalid or corrupt."):
    preprocessor.process_file(file_path)

def test_preprocessor_normalize_data(preprocessor):
  num_frames = 1 + int(preprocessor.config.size / preprocessor.config.hop)
  mfcc_size = preprocessor.config.nmfcc
  x = 2 * torch.ones((num_frames, mfcc_size))
  mean_expected = torch.ones((num_frames, mfcc_size))
  std_expected = 2 * torch.ones((num_frames, mfcc_size))
  y_expected = 0.5 * torch.ones((num_frames, mfcc_size))

  preprocessor.mean = mean_expected
  preprocessor.std = std_expected
  y = preprocessor.normalize_data(x)
  assert(torch.equal(preprocessor.mean, mean_expected)), "Invalid mean value."
  assert(torch.equal(preprocessor.std, std_expected)), "Invalid standard deviation value."
  assert(torch.equal(y, y_expected)), "Invalid normalized signal."
